NCalc v4 - Use Parlot instead of Antlr

gumbarros opened the pull request:
Closes #137
Closes #138 
Closes #141 
Closes #143

gumbarros commented:
Thanks everyone for the help at #137 :) 16 tests remaining now 🚀!

Bykiev commented on src/NCalc/Parser/LogicalExpressionParser.cs:
```diff
@@ -0,0 +1,254 @@
+        // The Deferred helper creates a parser that can be referenced by others before it is defined
+        var expression = Deferred<LogicalExpression>();
+
+        var intParser = Terms.Integer().Then<LogicalExpression>(static d => new ValueExpression(d));
```
We should pass `NumberOptions.AllowSign` to handle plus/minus sign

Bykiev commented on src/NCalc/Parser/LogicalExpressionParser.cs:
```diff
@@ -0,0 +1,254 @@
+            .Then<LogicalExpression>((context, x) =>
+                new ValueExpression(((LogicalExpressionParserContext)context).UseDecimalsAsDefault
+                    ? Convert.ToDecimal(x, CultureInfo.InvariantCulture)
+                    : Convert.ToDouble(x, CultureInfo.InvariantCulture)));
```
I think it can be combined into one with `ZeroOrOne`:

```
var decimalPointParser = Terms.Integer(NumberOptions.AllowSign)
    .AndSkip(Terms.Char('.'))
    .And(ZeroOrOne(Terms.Integer()))
    .Then<LogicalExpression>((context, x) =>
        new ValueExpression(((LogicalExpressionParserContext)context).UseDecimalsAsDefault
            ? Convert.ToDecimal(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture)
            : Convert.ToDouble(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture)));
```

Bykiev replied:
We should also handle scientific notification here, but it seems quite complicated

Bykiev replied:
Seems working fine (13 failing tests)

```
  var intParser = Terms.Integer(NumberOptions.AllowSign)
      .AndSkip(ZeroOrOne(Terms.Text("e", true)))
      .And(ZeroOrOne(Terms.Integer(NumberOptions.AllowSign)))
      .Then<LogicalExpression>((context, d) =>
      {
          if (d.Item2 != 0)
          {
              bool useDecimalAsDefault = ((LogicalExpressionParserContext)context).UseDecimalsAsDefault;

              if (useDecimalAsDefault)
              {
                  decimal val = Convert.ToDecimal(d.Item1 + "e" + d.Item2);
                  return new ValueExpression(val);
              }
              else
              {
                  double val = Convert.ToDouble(d.Item1 + "e" + d.Item2);
                  return new ValueExpression(val);
              }
          }

          return new ValueExpression(d.Item1);
      });

  var decimalPointParser = ZeroOrOne(Terms.Integer(NumberOptions.AllowSign))
      .AndSkip(Terms.Char('.'))
      .And(ZeroOrOne(Terms.Integer()))
      .AndSkip(ZeroOrOne(Terms.Text("e", true)))
      .And(ZeroOrOne(Terms.Integer(NumberOptions.AllowSign)))
      .Then<LogicalExpression>((context, x) =>
      {
          bool useDecimalAsDefault = ((LogicalExpressionParserContext)context).UseDecimalsAsDefault;

          if (useDecimalAsDefault)
          {
              decimal decimalValue = 0;
              if (x.Item3 != 0)
                  decimalValue = Convert.ToDecimal(x.Item1 + "." + x.Item2 + "e" + x.Item3, CultureInfo.InvariantCulture);
              else
                  decimalValue = Convert.ToDecimal(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture);

              return new ValueExpression(decimalValue);
          }
          else
          {
              double doubleValue = 0;
              if (x.Item3 != 0)
                  doubleValue = Convert.ToDouble(x.Item1 + "." + x.Item2 + "e" + x.Item3, CultureInfo.InvariantCulture);
              else
                  doubleValue = Convert.ToDouble(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture);

              return new ValueExpression(doubleValue);
          }
      });
```

Bykiev replied:
12 failing tests:

  ```
var intParser = Terms.Integer().Then<LogicalExpression>(static d => new ValueExpression(d));

  var intExponentParser = Terms.Integer(NumberOptions.AllowSign)
      .And(Terms.Text("e", true))
      .And(Terms.Integer(NumberOptions.AllowSign))
      .Then<LogicalExpression>((context, d) =>
      {
          if (!string.IsNullOrWhiteSpace(d.Item2))
          {
              bool useDecimalAsDefault = ((LogicalExpressionParserContext)context).UseDecimalsAsDefault;

              if (useDecimalAsDefault)
              {
                  decimal val = Convert.ToDecimal(d.Item1 + d.Item2 + d.Item3);
                  return new ValueExpression(val);
              }
              else
              {
                  double val = Convert.ToDouble(d.Item1 + d.Item2 + d.Item3);
                  return new ValueExpression(val);
              }
          }

          return new ValueExpression(d.Item1);
      });

  var decimalPointParser = ZeroOrOne(Terms.Integer(NumberOptions.AllowSign))
      .AndSkip(Terms.Char('.'))
      .And(ZeroOrOne(Terms.Integer()))
      .AndSkip(ZeroOrOne(Terms.Text("e", true)))
      .And(ZeroOrOne(Terms.Integer(NumberOptions.AllowSign)))
      .Then<LogicalExpression>((context, x) =>
      {
          bool useDecimalAsDefault = ((LogicalExpressionParserContext)context).UseDecimalsAsDefault;

          if (useDecimalAsDefault)
          {
              decimal decimalValue = 0;
              if (x.Item3 != 0)
                  decimalValue = Convert.ToDecimal(x.Item1 + "." + x.Item2 + "e" + x.Item3, CultureInfo.InvariantCulture);
              else
                  decimalValue = Convert.ToDecimal(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture);

              return new ValueExpression(decimalValue);
          }
          else
          {
              double doubleValue = 0;
              if (x.Item3 != 0)
                  doubleValue = Convert.ToDouble(x.Item1 + "." + x.Item2 + "e" + x.Item3, CultureInfo.InvariantCulture);
              else
                  doubleValue = Convert.ToDouble(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture);

              return new ValueExpression(doubleValue);
          }
      });

  var number = OneOf(
      decimalPointParser,
      intExponentParser,
      intParser
  );
```

Bykiev commented:
Some of the failing test are now working, don't know should we handle it. For example, ". + 2" was invalid expression but now returns result 2 (treats dot as 0)

gumbarros commented:
Thanks @Bykiev, with your changes only 12 tests remaining, mostly ternary operations and exception handling scenarios, I think we are in the endgame now. 

I will also document the breaking changes that I made for a cleaner codebase for v2.

Bykiev commented:
> Thanks @Bykiev, with your changes only 12 tests remaining, mostly ternary operations and exception handling scenarios, I think we are in the endgame now.

As I can see parser is still not fully implemented, a lot of BinaryExpressionType operators are missing. Also error handling is missing and I think it should me more robust

Bykiev commented on src/NCalc/Parser/LogicalExpressionParser.cs:
```diff
@@ -0,0 +1,301 @@
+        // The Recursive helper allows to create parsers that depend on themselves.
+        // ( "-" | "not" ) unary | primary;
+        var unary = Recursive<LogicalExpression>(u =>
+            minus.Or(Terms.Text("not")).And(u)
+                .Then<LogicalExpression>(static x =>
+                {
+                    return x.Item1 switch
+                    {
+                        "-" => new UnaryExpression(UnaryExpressionType.Negate, x.Item2),
+                        "not" => new UnaryExpression(UnaryExpressionType.Not, x.Item2),
+                        _ => throw new NotSupportedException()
+                    };
```
We should use case-insensitive comparision here

Bykiev commented on src/NCalc/Parser/LogicalExpressionParser.cs:
```diff
@@ -0,0 +1,301 @@
+            .And(Terms.Integer(NumberOptions.AllowSign))
+            .Then<LogicalExpression>((context, d) =>
+            {
+                if (!string.IsNullOrWhiteSpace(d.Item2))
```
I've reviewed it and if `(!string.IsNullOrWhiteSpace(d.Item2))` condition should be removed as we're using `And` and this one will be triggered only when it's specified

Bykiev commented:
> Some of the failing test are now working, don't know should we handle it. For example, ". + 2" was invalid expression but now returns result 2 (treats dot as 0)

We can use smth like this to achieve it:

```
  Terms.Integer(NumberOptions.AllowSign).Then<long?>(x => x).Or(Empty<long?>(null))
  .AndSkip(Terms.Char('.'))
  .And(Terms.Integer().Then<long?>(x => x).Or(Empty<long?>(null)))
  .And(ZeroOrOne(Terms.Text("e", true)))
  .And(ZeroOrOne(Terms.Integer(NumberOptions.AllowSign)))
```
if x.Item1 == null and x.Item2 == null then we can throw an error

gumbarros commented:
9 tests remaining 🙏

Help at `ShouldHandleOperatorsPriority` is appreciated, stuck in how to execute things inside () before anything. After this, I think we are missing just ternary operator and left and right shift operators

Bykiev commented on src/NCalc/Parser/LogicalExpressionParser.cs:
```diff
@@ -0,0 +1,348 @@
+                var result = x.Item1;
+                foreach (var op in x.Item2)
+                {
+                    result = op.Item1.ToLowerInvariant() switch
```
`ToUpperInvariant` is a preferred way

Bykiev commented:
> 9 tests remaining 🙏
> 
> Help at `ShouldHandleOperatorsPriority` is appreciated, stuck in how to execute things inside () before anything. After this, I think we are missing just ternary operator and left and right shift operators

I'm not 100% sure, but looks like unary parser should be after binary operators

Bykiev commented:
7 tests failing. I can't contribute to this branch, that's why I forked your repo, here is a changes: https://github.com/Bykiev/ncalc/commit/798615b14761470be9714438e54294fb63e763d4

Bykiev commented:
6 failing tests: https://github.com/Bykiev/ncalc/commit/cb98f369196c665dba499f6e933458ee424802a5

Bykiev commented:
Fix mistype https://github.com/Bykiev/ncalc/commit/b7d6e9be8a389b4ae21e4a04a6fc2a500e7f189d

gumbarros commented:
@Bykiev tks, merged via PR at my repo.

Bykiev commented:
> @Bykiev tks, merged via PR at my repo.

No problem, we need to implement left/right shifts, research the issue with NOT casing (breaks some tests, but fixes another) and handle errors. 
It's time to sleep...

gumbarros commented:
![image](https://github.com/ncalc/ncalc/assets/52143624/c21a30b1-69da-419a-b21e-894fbd09bf69)
The error of NOT is because not is being identified as a Identifier and not a unary operator. I'm trying to fix this, but no success, maybe tomorrow, will keep updates here :)

gumbarros commented:
4 tests remaining!

Remaining items:

- [ ] Syntax errors
- [ ] Ternary operator
- [ ] Right and left shift
- [ ] NOT as unary instead of identifier

Bykiev commented:
> ![image](https://private-user-images.githubusercontent.com/52143624/331747071-c21a30b1-69da-419a-b21e-894fbd09bf69.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYwMjYxODEsIm5iZiI6MTcxNjAyNTg4MSwicGF0aCI6Ii81MjE0MzYyNC8zMzE3NDcwNzEtYzIxYTMwYjEtNjlkYS00MTlhLWIyMWUtODk0ZmJkMDliZjY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE4VDA5NTEyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiYTQyNzk4ZGRkMGZlYWFkN2QwNzZiZjc4ZTk3OWMyY2Y0MDJlNDZhNzkzMTdkMjBkNTNlOGZmYTZlMzEwMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.u2yIfR0H_-zPAm9PgqiiPoqTaQvyMd4UGeVXJy-kj1A) The error of "NOT" is because "NOT" is being identified as a `Identifier` and not a unary operator. I'm trying to fix this, but no success, maybe tomorrow, will keep updates here :)

Can't understand, why does AND/OR are not treated as identifier?

Bykiev commented:
@sebastienros, can you please look at it?

Bykiev commented:
3 test remaining: https://github.com/Bykiev/ncalc/commit/168e5a3af558fda5ed5053f36f71f2752ca08a50

gumbarros commented:
I created `Should_Evaluate_Not_Unary_Operator` very stuck on this, sounds like a Parlot bug, when I change "NOT" to "not" a test works and another fails.

Bykiev commented:
> I created `Should_Evaluate_Not_Unary_Operator` very stuck on this, sounds like a Parlot bug, when I change "NOT" to "not" a test works and another fails.

Waiting for @sebastienros response, indeed looking like a bug

gumbarros commented:
@Bykiev weird behavior, but solved NOT bug by using `Literal` instead of `Term`, discovered about `Literal` in a [Sebastien comment here](https://github.com/sebastienros/parlot/issues/108)

Bykiev commented:
I've tested your repo and not still failing for me event with Literals

gumbarros commented:
@Bykiev I moved the tests to xUnit and now we have 10 failing tests, BUT, these were already failing on the older version, it's now easier to see what expressions are failing because of I migrated many expressions to `TheoryAttribute` and `InlineDataAttribute`.

![image](https://github.com/ncalc/ncalc/assets/52143624/17aa43db-6b2f-4d14-a8c1-c968856f50e5)

It's now more easier to debug because it's possible to select what expression you want to debug

gumbarros commented:
@sebastienros 
![image](https://github.com/ncalc/ncalc/assets/52143624/eef39487-01ef-431e-a7ab-170274b12587)

Help is appreciated in this specific unit test, I'm trying multiple things, but does not makes sense to me that "NOT" is parsed as unary and "not" is not ☠️

gumbarros commented:
![image](https://github.com/ncalc/ncalc/assets/52143624/09bcf0f1-c19b-47e9-bf90-9d0ac5e85220)
I give up, time to some Sunday rest, does not make sense for me some working and some fail

Bykiev commented:
I can't run tests after migration to xunit, also it seems we can't run tests against .net framework?

gumbarros commented:
> I can't run tests after migration to xunit, also it seems we can't run tests against .net framework?

Here working normally. I removed net48 temporaly to see how many tests because they were running 2x in each framework, I added again for you.

Try cleaning your solution and running again, strange.

Bykiev commented:
I was able to make it working with some fixes in project files

Bykiev commented:
I'm pretty sure it's a Parlot bug, let's wait for Sebastien response, it seems he is busy now

sebastienros commented:
I will look at why the expression tests are failing. I see some other refactoring to do to simplify the grammar.

sebastienros commented:
Bug identified, fix incoming. `OneOf` is able to build a lookup table if all its parsers expect a specific char. So in the case of `unary`, it can build one. But for `"not"` it return `n` as the first expected char, even though it was case insensitive. Then the parser would fail and `NOT` would be parsed as an identifier.

gumbarros commented:
> Bug identified, fix incoming. `OneOf` is able to build a lookup table if all its parsers expect a specific char. So in the case of `unary`, it can build one. But for `"not"` it return `n` as the first expected char, even though it was case insensitive. Then the parser would fail and `NOT` would be parsed as an identifier.

Thank you very much, looking forward to update the NuGet when available

gumbarros commented:
![image](https://github.com/ncalc/ncalc/assets/52143624/77a42b72-bdc3-4320-989e-f6de352c5910)
Public API now available via DocFX :)
I will create the CI to publish to GitHub pages after this branch is merged.

gumbarros commented:
![image](https://github.com/ncalc/ncalc/assets/52143624/79160faa-aac7-4093-a9b6-eaf770df2607)

@Bykiev all tests now working! If you want to review or make some changes, you will need to manually clone Parlot at `..\..\src\NCalc\NCalc.csproj`

@sebastienros thanks again for your work at Parlot, can you make a NuGet release when possible? 

I think tomorrow is a good date for a NCalc release if everything is alright with this PR :)

sebastienros commented:
0.0.27 on nuget.

Here is what I worked on locally. I think I broken something at some point while working on a specific bug, but there are multiple constructs you could try to copy while ensuring it's still working. It's making the grammar simpler to read, and faster too.

```c#
#nullable disable

// ReSharper disable AssignNullToNotNullAttribute

using NCalc.Domain;
using Parlot.Fluent;
using static Parlot.Fluent.Parsers;
using Identifier = NCalc.Domain.Identifier;

namespace NCalc.Parser;

public static class LogicalExpressionParser
{
    private static readonly Parser<LogicalExpression> Parser;

    private static readonly ValueExpression True = new(true);
    private static readonly ValueExpression False = new(false);

    static LogicalExpressionParser()
    {
        /*
         * Grammar:
         * expression     => ternary ( ( "-" | "+" ) ternary )* ;
         * ternary        => logical ( "?" logical ":" logical)?
         * logical        => equality ( ( "and" | "or" ) equality )* ;
         * equality       => relational ( ( "=" | "!=" | ... ) relational )* ;
         * relational     => shift ( ( ">=" | ">" | ... ) shift )* ;
         * shift          => additive ( ( "<<" | ">>" ) additive )* ;
         * additive       => multiplicative ( ( "-" | "+" ) multiplicative )* ;
         * multiplicative => exponential ( "/" | "*" | "%") exponential )* ;
         * exponential    => unary ( "**" ) unary )* ;
         * unary          => ( "-" | "not" | "!" ) primary
         *
         * primary        => NUMBER
         *                  | STRING
         *                  | "true"
         *                  | "false"
         *                  | "[" anything "]"
         *                  | function
         *                  | "(" expression ")" ;
         *
         * function       => Identifier "(" arguments ")"
         * arguments      => expression ( "," expression )*
         */
        // The Deferred helper creates a parser that can be referenced by others before it is defined
        var expression = Deferred<LogicalExpression>();

        var exponentNumberPart = Literals.Text("e", true).SkipAnd(Literals.Integer(NumberOptions.AllowSign)).Then(x => x);
        var decimalNumberPart = Literals.Char('.').SkipAnd(Terms.Integer());
        var integralNumberPart = Literals.Integer(NumberOptions.AllowSign);

        // [integral_value]['.'decimal_value}]['e'exponent_value]
        var number =
            SkipWhiteSpace(OneOf(
                Always<long>(0).And(decimalNumberPart.Then<long?>(x => x)).And(ZeroOrOne(exponentNumberPart).ThenElse<long?>(x => x, null)).Then(x => (x.Item1, x.Item2, x.Item3)),
                integralNumberPart.And(ZeroOrOne(decimalNumberPart).ThenElse<long?>(x => x, null)).And(ZeroOrOne(exponentNumberPart).ThenElse<long?>(x => x, null)).Then(x => (x.Item1, x.Item2, x.Item3))
                ))
            .Then<LogicalExpression>((ctx, x) =>
            {
                long integralValue = x.Item1;
                long? decimalPart = x.Item2;
                long? exponentPart = x.Item3;

                decimal result = integralValue;

                // decimal part?
                if (decimalPart != null && decimalPart.Value != 0)
                {
                    var digits = Math.Floor(Math.Log10(decimalPart.Value) + 1);
                    result = result + decimalPart.Value / (decimal)Math.Pow(10, digits);
                }

                // exponent part?
                if (exponentPart != null)
                {
                    result = result * (decimal)Math.Pow(10, exponentPart.Value);
                }

                if (ctx is LogicalExpressionParserContext parseContext && parseContext.UseDecimalsAsDefault)
                {
                    return new ValueExpression(result);
                }
                else
                {
                    if (decimalPart != null || (exponentPart != null))
                    {
                        return new ValueExpression((double)result);
                    }
                    else
                    {
                        return new ValueExpression((long)result);
                    }
                }                
            });
        
        var comma = Terms.Char(',');
        var divided = Terms.Char('/');
        var times = Terms.Char('*');
        var modulo = Terms.Char('%');
        var minus = Terms.Text("-");
        var plus = Terms.Text("+");
        var bitwiseNot = Terms.Text("~");
        var exponent = Terms.Text("**");
        var openParen = Terms.Char('(');
        var closeParen = Terms.Char(')');
        var openBrace = Terms.Char('[');
        var closeBrace = Terms.Char(']');
        var questionMark = Terms.Char('?');
        var colon = Terms.Char(':');
        
        var negate = Terms.Text("!");
        var not = Terms.Text("not", true);

        // "(" expression ")"
        var groupExpression = Between(openParen, expression, closeParen.ElseError("Parenthesis not closed."));

        // "[" identifier "]"
        var identifierExpression = openBrace
            .SkipAnd(AnyCharBefore(closeBrace, consumeDelimiter: true))
            .Or(Terms.Identifier())
            .Then<LogicalExpression>(x => new Identifier(x.ToString()));

        var arguments = Separated(comma, expression);

        var functionWithArguments = Terms
            .Identifier()
            .And(openParen.SkipAnd(arguments).AndSkip(closeParen))
            .Then<LogicalExpression>(x => new Function(new Identifier(x.Item1.ToString()), x.Item2.ToArray()));

        var functionWithoutArguments = Terms
            .Identifier()
            .And(openParen.AndSkip(closeParen))
            .Then<LogicalExpression>(x => new Function(new Identifier(x.Item1.ToString()), []));

        var function = OneOf(functionWithArguments, functionWithoutArguments);
        
        var booleanTrue = Terms.Text("true", true)
            .Then<LogicalExpression>(True);
        var booleanFalse = Terms.Text("false", true)
            .Then<LogicalExpression>(False);
        
        var stringValue = Terms.String(quotes: StringLiteralQuotes.SingleOrDouble)
            .Then<LogicalExpression>(x => new ValueExpression(x.ToString()));

        var charIsNumber = Literals.Pattern(char.IsNumber);

        var dateTimeParser = Terms
            .Char('#')
            .SkipAnd(charIsNumber)
            .AndSkip(divided)
            .And(charIsNumber)
            .AndSkip(divided)
            .And(charIsNumber)
            .AndSkip(Literals.Char('#'))
            .Then<LogicalExpression>(date =>
            {
                if (DateTime.TryParse($"{date.Item1}/{date.Item2}/{date.Item3}", out var result))
                {
                    return new ValueExpression(result);
                }

                throw new FormatException("Invalid date format.");
            });

        // primary => NUMBER | "[" identifier "]" | DateTime | string | function | boolean | "(" expression ")";
        
        var primary = OneOf(
            number, 
            booleanTrue, 
            booleanFalse,
            dateTimeParser,
            stringValue, 
            function,
            groupExpression,
            identifierExpression);

        // exponential => unary ( "**" unary )* ;
        var exponential = primary.And(ZeroOrMany(exponent.And(primary)))
            .Then(static x =>
            {
                LogicalExpression result = null;

                if (x.Item2.Count == 0)
                {
                    result = x.Item1;
                }
                else if (x.Item2.Count == 1)
                {
                    result = new BinaryExpression(BinaryExpressionType.Exponentiation, x.Item1, x.Item2[0].Item2);
                }
                else
                {
                    for (int i = x.Item2.Count - 1; i > 0; i--)
                    {
                        result = new BinaryExpression(BinaryExpressionType.Exponentiation, x.Item2[i - 1].Item2, x.Item2[i].Item2);
                    }

                    result = new BinaryExpression(BinaryExpressionType.Exponentiation, x.Item1, result);
                }

                return result;
            });

        // The Recursive helper allows to create parsers that depend on themselves.
        // ( "-" | "not" ) unary | primary;
        var unary = Recursive<LogicalExpression>(u =>
                OneOf(
                    not.Then(UnaryExpressionType.Not),
                    negate.Then(UnaryExpressionType.Not),
                    minus.Then(UnaryExpressionType.Negate),
                    bitwiseNot.Then(UnaryExpressionType.BitwiseNot))
                .And(u).Then<LogicalExpression>(static x => new UnaryExpression(x.Item1, x.Item2))
                .Or(exponential)
                );
        
        // multiplicative => unary ( ( "/" | "*" | "%" ) unary )* ;
        var multiplicative = unary.And(ZeroOrMany(
            divided.Then(BinaryExpressionType.Div)
            .Or(times.Then(BinaryExpressionType.Times))
            .Or(modulo.Then(BinaryExpressionType.Modulo))
            .And(unary)))
            .Then(static x =>
            {
                var result = x.Item1;

                foreach (var op in x.Item2)
                {
                    result = new BinaryExpression(op.Item1, result, op.Item2);
                }

                return result;
            }).Or(unary);

        // additive => multiplicative ( ( "-" | "+" ) multiplicative )* ;
        var additive = multiplicative.And(ZeroOrMany(
            plus.Then(BinaryExpressionType.Plus)
            .Or(minus.Then(BinaryExpressionType.Minus))
            .And(multiplicative)))
            .Then(static x =>
            {
                var result = x.Item1;
                foreach (var op in x.Item2)
                {
                    result = new BinaryExpression(op.Item1, result, op.Item2);
                }

                return result;
            }).Or(multiplicative);

        // shift => additive ( ( "<<" | ">>" ) additive )* ;
        var shift = additive.And(ZeroOrMany(
                Terms.Text("<<").Then(BinaryExpressionType.LeftShift)
                .Or(Terms.Text(">>").Then(BinaryExpressionType.RightShift))
                .And(additive)))
            .Then(static x =>
            {
                var result = x.Item1;
                foreach (var op in x.Item2)
                {
                    result = new BinaryExpression(op.Item1, result, op.Item2);
                }

                return result;
            }).Or(additive);

        // relational => shift ( ( ">=" | "<=" | "<" | ">" ) shift )* ;
        var relational = shift.And(ZeroOrMany(OneOf(
                    Terms.Text(">=").Then(BinaryExpressionType.GreaterOrEqual),
                    Terms.Text("<=").Then(BinaryExpressionType.LesserOrEqual),
                    Terms.Text("<").Then(BinaryExpressionType.Lesser),
                    Terms.Text(">").Then(BinaryExpressionType.Greater))
                .And(shift)))
            .Then(static x =>
            {
                var result = x.Item1;
                foreach (var op in x.Item2)
                {
                    result = new BinaryExpression(op.Item1, result, op.Item2);
                }

                return result;
            });

        var equality = relational.And(ZeroOrMany(OneOf(
                    Terms.Text("<>").Then(BinaryExpressionType.NotEqual),
                    Terms.Text("==").Then(BinaryExpressionType.Equal),
                    Terms.Text("!=").Then(BinaryExpressionType.NotEqual),
                    Terms.Text("=").Then(BinaryExpressionType.Equal))
                .And(relational)))
            .Then(static x =>
            {
                var result = x.Item1;
                foreach (var op in x.Item2)
                {
                    result = new BinaryExpression(op.Item1, result, op.Item2);
                }

                return result;
            });

        var and = Terms
            .Text("AND", true).Then(BinaryExpressionType.And)
            .Or(Terms.Text("&&").Then(BinaryExpressionType.And))
            .Or(Terms.Text("&").Then(BinaryExpressionType.BitwiseAnd));

        var or = Terms
            .Text("OR", true).Then(BinaryExpressionType.Or)
            .Or(Terms.Text("||").Then(BinaryExpressionType.Or))
            .Or(Terms.Text("|").Then(BinaryExpressionType.BitwiseOr));

        var xor = Terms.Text("^").Then(BinaryExpressionType.BitwiseXOr);

        // logical => equality ( ( "and" | "or" ) equality )* ;
        var logical = equality.And(ZeroOrMany(OneOf(and, or, xor).And(equality)))
            .Then(static x =>
            {
                var result = x.Item1;
                foreach (var op in x.Item2)
                {
                    result = new BinaryExpression(op.Item1, result, op.Item2);
                }

                return result;
            });

        // ternary => logical("?" logical ":" logical) ?
        var ternary = logical.And(ZeroOrOne(questionMark.SkipAnd(logical).AndSkip(colon).And(logical)))
            .Then(x => x.Item2.Item1 == null
                ? x.Item1
                : new TernaryExpression(x.Item1, x.Item2.Item1, x.Item2.Item2));

        expression.Parser = ternary;
        Parser = expression;
    }

    public static LogicalExpression Parse(LogicalExpressionParserContext context) => Parser.Parse(context);
}

```

sebastienros commented:
I found the problem in the parser I pasted, it's fixed now. All tests pass minus the one you disabled.

gumbarros commented:
> I found the problem in the parser I pasted, it's fixed now. All tests pass minus the one you disabled.

Tks, everything alright on net8.0, but on 4.8 the following exception is throw:

```cs
Could not load file or assembly 'System.ValueTuple, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified
```

sebastienros commented:
can you try to target 462 instead of 461 ?

sebastienros commented:
NB: 461 is not supported anymore. 4.8.1 is installed on all machines with windows update. 4.6.2 can be installed manually. That gives you an idea of which ones to actually support/target.

gumbarros commented:
> can you try to target 462 instead of 461 ?
I'm targeting `net48` at the unit test

sebastienros commented:
You can also add a reference to <PackageReference Include="System.ValueTuple" Version="4.5.0" /> in NCalc for the net46x specifically.

gumbarros commented:
![image](https://github.com/ncalc/ncalc/assets/52143624/47238ab4-8cd2-47f1-9042-8bf8d83b7d20)

Targeting net48 still causes errors, just standard works, people at older .NET framework versions can just target standard right?

sebastienros commented:
This could be an issue only for those on 4.6.2 still, see https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting

But anyone above 4.7.1 will be fine.

gumbarros commented:
> This could be an issue only for those on 4.6.2 still, see https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting
> 
> But anyone above 4.7.1 will be fine.

Good point, re-added the support using your suggested conditional at .csproj

gumbarros commented:
@Bykiev after improving `Should_Detect_Syntax_Errors_Before_Evaluation` and your review I think we are ready :)

Bykiev commented:
Good job guys, I'm going to add a benchmark project and review the code at the evening (I'm on +3 UTC)

Bykiev commented:
@gumbarros, I've added a benchmark project and simple tests, feel free to add new one

Bykiev commented:
It seems results with `ANTLR` and `Parlot` are very close:

```
// * Summary *

BenchmarkDotNet v0.13.12, Windows 10 (10.0.19044.2965/21H2/November2021Update)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.205
  [Host] : .NET 6.0.30 (6.0.3024.21525), X64 RyuJIT AVX2

Job=ShortRun  IterationCount=3  LaunchCount=1
WarmupCount=3
```

| Method                 | Runtime              | NuGetReferences | Mean       | Error       | StdDev    |
|----------------------- |--------------------- |---------------- |-----------:|------------:|----------:|
| SimpleExpression       | .NET 8.0             | Default         |   937.7 ns |   546.80 ns |  29.97 ns |
| EvaluateCustomFunction | .NET 8.0             | Default         |   296.1 ns |    84.40 ns |   4.63 ns |
| EvaluateParameters     | .NET 8.0             | Default         | 1,451.7 ns |    13.85 ns |   0.76 ns |
| SimpleExpression       | .NET 8.0             | NCalcSync       |   945.9 ns |   344.83 ns |  18.90 ns |
| EvaluateCustomFunction | .NET 8.0             | NCalcSync       |   364.8 ns |    66.49 ns |   3.64 ns |
| EvaluateParameters     | .NET 8.0             | NCalcSync       | 2,052.7 ns | 2,826.09 ns | 154.91 ns |
| SimpleExpression       | .NET Framework 4.6.2 | Default         | 1,083.1 ns | 2,581.25 ns | 141.49 ns |
| EvaluateCustomFunction | .NET Framework 4.6.2 | Default         |   394.4 ns | 1,590.46 ns |  87.18 ns |
| EvaluateParameters     | .NET Framework 4.6.2 | Default         | 1,523.2 ns |   452.64 ns |  24.81 ns |
| SimpleExpression       | .NET Framework 4.6.2 | NCalcSync       |   915.8 ns |    77.51 ns |   4.25 ns |
| EvaluateCustomFunction | .NET Framework 4.6.2 | NCalcSync       |   295.9 ns |   198.88 ns |  10.90 ns |
| EvaluateParameters     | .NET Framework 4.6.2 | NCalcSync       | 1,439.5 ns |   330.80 ns |  18.13 ns |

gumbarros commented:
@Bykiev I think something wrong in the benchmark, the memory usage is the same even with Cache disabled.


```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3593/23H2/2023Update/SunValley3)
13th Gen Intel Core i5-1335U, 1 CPU, 12 logical and 10 physical cores
.NET SDK 8.0.204
  [Host] : .NET 6.0.29 (6.0.2924.17105), X64 RyuJIT AVX2

Job=ShortRun  IterationCount=3  LaunchCount=1
WarmupCount=3
```

| Method                 | Runtime              | NuGetReferences | Mean      | Error     | StdDev    | Rank | Gen0   | Gen1   | Allocated |
|----------------------- |--------------------- |---------------- |----------:|----------:|----------:|-----:|-------:|-------:|----------:|
| EvaluateCustomFunction | .NET 8.0             | Default         |  5.290 us |  2.668 us | 0.1462 us |    1 | 0.1984 | 0.0076 |   1.23 KB |
| EvaluateCustomFunction | .NET Framework 4.6.2 | Default         |  8.687 us |  3.372 us | 0.1848 us |    2 | 0.1984 |      - |   1.23 KB |
| EvaluateCustomFunction | .NET 8.0             | NCalcSync       |  8.701 us |  8.588 us | 0.4707 us |    2 | 0.1984 | 0.0076 |   1.23 KB |
| EvaluateCustomFunction | .NET Framework 4.6.2 | NCalcSync       |  8.917 us |  1.426 us | 0.0782 us |    3 | 0.1984 |      - |   1.23 KB |
| SimpleExpression       | .NET 8.0             | Default         | 15.948 us |  2.442 us | 0.1339 us |    4 | 0.7629 |      - |   4.69 KB |
| EvaluateParameters     | .NET 8.0             | Default         | 17.164 us |  8.992 us | 0.4929 us |    5 | 0.7019 | 0.0305 |   4.43 KB |
| SimpleExpression       | .NET 8.0             | NCalcSync       | 20.668 us | 44.476 us | 2.4379 us |    6 | 0.7629 | 0.0305 |   4.69 KB |
| SimpleExpression       | .NET Framework 4.6.2 | Default         | 23.854 us |  8.551 us | 0.4687 us |    7 | 0.7629 | 0.0305 |   4.69 KB |
| EvaluateParameters     | .NET Framework 4.6.2 | Default         | 25.517 us | 28.141 us | 1.5425 us |    8 | 0.7019 | 0.0305 |   4.43 KB |
| EvaluateParameters     | .NET Framework 4.6.2 | NCalcSync       | 29.958 us | 61.970 us | 3.3968 us |    9 | 0.6714 | 0.0610 |   4.43 KB |
| EvaluateParameters     | .NET 8.0             | NCalcSync       | 31.624 us | 63.700 us | 3.4916 us |   10 | 0.7019 | 0.0305 |   4.43 KB |
| SimpleExpression       | .NET Framework 4.6.2 | NCalcSync       | 32.799 us | 91.086 us | 4.9927 us |   11 | 0.7629 | 0.0305 |   4.69 KB |


And this breaking change should not compile at the benchmark:

```        [Benchmark]
        public void SimpleExpression()
        {
            string expression = "1 - ( 3 + 2.5 ) * 4 - 1 / 2 + 1 - ( 3 + 2.5 ) * 4 - 1 / 2 + 1 - ( 3 + 2.5 ) * 4 - 1 / 2";

            Expression.CacheEnabled = false;
            _ = LogicalExpressionFactory.Create("1+1"); //This should not work at NuGet version.
            var expr = new Expression(expression);
            expr.Evaluate();
        }
```

Bykiev commented:
Hmm, I'll look at it, doesn't see any issues

Bykiev commented:
I did 2 independent tests: 1st with current branch, 2nd - with NuGet and got even worse results:

// * Summary *

BenchmarkDotNet v0.13.12, Windows 10 (10.0.19044.2965/21H2/November2021Update)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.205
  [Host] : .NET 6.0.30 (6.0.3024.21525), X64 RyuJIT AVX2

Job=ShortRun  IterationCount=3  LaunchCount=1
WarmupCount=3

| Method           | Runtime              | Mean       | Error      | StdDev   |
|----------------- |--------------------- |-----------:|-----------:|---------:|
| SimpleExpression | .NET 8.0             | 1,019.9 ns |   703.7 ns | 38.57 ns |
| SimpleExpression | .NET Framework 4.6.2 |   954.2 ns | 1,008.9 ns | 55.30 ns |



// * Summary *

BenchmarkDotNet v0.13.12, Windows 10 (10.0.19044.2965/21H2/November2021Update)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.205
  [Host] : .NET 6.0.30 (6.0.3024.21525), X64 RyuJIT AVX2

Job=ShortRun  IterationCount=3  LaunchCount=1
WarmupCount=3

| Method           | Runtime              | Mean     | Error    | StdDev   |
|----------------- |--------------------- |---------:|---------:|---------:|
| SimpleExpression | .NET 8.0             | 890.1 ns | 597.5 ns | 32.75 ns |
| SimpleExpression | .NET Framework 4.6.2 | 845.2 ns | 186.2 ns | 10.20 ns |

gumbarros commented:
Are you sure that's really NuGet version? Because
```cs
   _ = LogicalExpressionFactory.Create("1+1"); //This should not work at NuGet version.
   ```

Bykiev commented:
> Are you sure that's really NuGet version? Because
> 
> ```cs
>    _ = LogicalExpressionFactory.Create("1+1"); //This should not work at NuGet version.
> ```

I'm 100% sure, my test doesn't use `LogicalExpressionFactory`. Don't know how it's working internally in BenchmarkDotNet, but you can view code with ILSpy

gumbarros commented:
I manually added NuGet to the benchmark and obtained different memory usage:

Memory usage without NuGet:
| Method                 | Runtime              | NuGetReferences  | Mean      | Error    | StdDev    | Rank | Gen0   | Allocated |
|----------------------- |--------------------- |----------------- |----------:|---------:|----------:|-----:|-------:|----------:|
| EvaluateCustomFunction | .NET 8.0             | Default          |  3.861 us | 1.623 us | 0.0889 us |    1 | 0.1984 |   1.23 KB |
| EvaluateCustomFunction | .NET 8.0             | NCalcSync.signed |  3.925 us | 1.180 us | 0.0647 us |    2 | 0.1984 |   1.23 KB |
| EvaluateCustomFunction | .NET Framework 4.6.2 | NCalcSync.signed |  4.081 us | 1.724 us | 0.0945 us |    3 | 0.1984 |   1.23 KB |
| EvaluateCustomFunction | .NET Framework 4.6.2 | Default          |  4.178 us | 1.105 us | 0.0606 us |    4 | 0.1984 |   1.23 KB |
| SimpleExpression       | .NET 8.0             | Default          | 10.300 us | 3.099 us | 0.1699 us |    5 | 0.6866 |   4.25 KB |
| SimpleExpression       | .NET 8.0             | NCalcSync.signed | 10.705 us | 4.950 us | 0.2713 us |    6 | 0.6866 |   4.25 KB |
| SimpleExpression       | .NET Framework 4.6.2 | NCalcSync.signed | 10.732 us | 1.616 us | 0.0886 us |    6 | 0.6866 |   4.25 KB |
| SimpleExpression       | .NET Framework 4.6.2 | Default          | 10.786 us | 1.881 us | 0.1031 us |    6 | 0.6866 |   4.25 KB |
| EvaluateParameters     | .NET 8.0             | Default          | 12.625 us | 7.373 us | 0.4041 us |    7 | 0.7172 |   4.43 KB |
| EvaluateParameters     | .NET Framework 4.6.2 | NCalcSync.signed | 12.872 us | 3.915 us | 0.2146 us |    8 | 0.7172 |   4.43 KB |
| EvaluateParameters     | .NET Framework 4.6.2 | Default          | 13.033 us | 1.451 us | 0.0795 us |    9 | 0.7172 |   4.43 KB |
| EvaluateParameters     | .NET 8.0             | NCalcSync.signed | 13.163 us | 3.651 us | 0.2001 us |    9 | 0.7172 |   4.43 KB |


Manually referenced NuGet:

| Method                 | Runtime              | Mean      | Error     | StdDev    | Rank | Gen0   | Gen1   | Allocated |
|----------------------- |--------------------- |----------:|----------:|----------:|-----:|-------:|-------:|----------:|
| EvaluateCustomFunction | .NET Framework 4.6.2 |  9.369 us |  3.466 us | 0.1900 us |    1 | 2.0142 | 0.0610 |  12.39 KB |
| EvaluateCustomFunction | .NET 8.0             |  9.463 us |  6.586 us | 0.3610 us |    1 | 2.0142 | 0.0610 |  12.39 KB |
| EvaluateParameters     | .NET 8.0             | 29.501 us |  6.730 us | 0.3689 us |    2 | 5.7983 | 0.3662 |  35.76 KB |
| EvaluateParameters     | .NET Framework 4.6.2 | 30.573 us |  5.675 us | 0.3111 us |    3 | 5.8289 | 0.3967 |  35.76 KB |
| SimpleExpression       | .NET 8.0             | 32.082 us | 12.637 us | 0.6927 us |    4 | 5.4321 | 0.4883 |  33.61 KB |
| SimpleExpression       | .NET Framework 4.6.2 | 33.387 us | 13.232 us | 0.7253 us |    5 | 5.4321 | 0.4883 |  33.61 KB |

gumbarros commented:
Maybe we can try:
https://stackoverflow.com/questions/58990324/how-to-benchmark-two-different-versions-of-the-same-non-nuget-library-in-benchma

Bykiev commented:
Docs from BenchmarkDotNet:

> InProcessEmitToolchain is our toolchain which does not generate any new executable. It emits IL on the fly and runs it from within the process itself. It can be useful if want to run the benchmarks very fast or if you want to run them for framework which we don't support. An example could be a local build of CoreCLR.

gumbarros commented:
@Bykiev what do you think?

 Method                                    | Runtime              | Mean      | Error     | StdDev    | Rank | Gen0   | Gen1   | Allocated |
|------------------------------------------ |--------------------- |----------:|----------:|----------:|-----:|-------:|-------:|----------:|
| CsProjExpression_EvaluateCustomFunction   | .NET 8.0             |  4.127 us |  2.851 us | 0.1563 us |    1 | 0.1984 |      - |   1.23 KB |
| CsProjExpression_EvaluateCustomFunction   | .NET Framework 4.6.2 |  4.263 us |  3.324 us | 0.1822 us |    2 | 0.1984 |      - |   1.23 KB |
| NuGetExpression_EvaluateCustomFunction    | .NET 8.0             |  9.695 us |  3.352 us | 0.1837 us |    3 | 2.0142 | 0.0610 |  12.39 KB |
| NuGetExpression_EvaluateCustomFunction    | .NET Framework 4.6.2 | 10.021 us |  4.609 us | 0.2527 us |    4 | 2.0142 | 0.0610 |  12.39 KB |
| CsProjExpression_EvaluateParameters       | .NET 8.0             | 12.972 us |  2.948 us | 0.1616 us |    5 | 0.7172 |      - |   4.43 KB |
| CsProjExpression_EvaluateParameters       | .NET Framework 4.6.2 | 13.019 us |  1.416 us | 0.0776 us |    5 | 0.7172 |      - |   4.43 KB |
| NuGetExpression_EvaluateParameters        | .NET Framework 4.6.2 | 30.976 us |  6.944 us | 0.3806 us |    6 | 5.8289 | 0.3967 |  35.76 KB |
| NuGetExpression_EvaluateParameters        | .NET 8.0             | 31.168 us |  6.049 us | 0.3316 us |    6 | 5.7983 | 0.3662 |  35.76 KB |
| NuGetNewVersionExpressionSimpleExpression | .NET Framework 4.6.2 | 34.583 us | 13.115 us | 0.7189 us |    7 | 5.4321 | 0.4883 |  33.61 KB |
| NuGetNewVersionExpressionSimpleExpression | .NET 8.0             | 34.665 us |  9.289 us | 0.5092 us |    7 | 5.4321 | 0.4883 |  33.61 KB |

Ready to merge?

Bykiev commented:
I still need more time to review all the changes. 
I can't build the benchmark project on my laptop, but your result looks more plausible. It seems it's a bug with InProcess toolchain in BenchmarkDotNet

gumbarros commented:
> I still need more time to review all the changes. I can't build the benchmark project on my laptop, but your result looks more plausible. It seems it's a bug with InProcess toolchain in BenchmarkDotNet

![image](https://github.com/ncalc/ncalc/assets/52143624/6cf7f1c4-291b-4a41-942a-3a9e76665189)

The benchmark only works in this config, Debug and Release causes a Assembly Name collision.

Bykiev commented:
It will be good to run benchmarks on CI, it can be done after release in a separate PR

gumbarros commented:
> It will be good to run benchmarks on CI, it can be done after release in a separate PR

Yes, my regret in this version was that I got carried away and ended up not creating a preview branch and creating several smaller PRs.

Bykiev commented:
I see some issues:
1. NCalc.Benchmarks depends on NCalsSync package, seems should be signed version
2. `extern alias NuGetVersion;` seems incorrect, should be 
```
using NuGetExpression = NCalc.Expression;
using NuGetDomain = NCalc;
```
3. Solution seems to be incorrect, 
```
{FC5C9A17-744F-4F97-B0FC-330B0514B395}.SignedRelease|Any CPU.ActiveCfg = Debug|Any CPU
{FC5C9A17-744F-4F97-B0FC-330B0514B395}.SignedRelease|Any CPU.Build.0 = Debug|Any CPU
```
it should be with Release

Bykiev commented:
Due to antivirus I can't run the benchmark project, but should be enough to run it on CI later

gumbarros commented:
> I see some issues:
> 
> 1. NCalc.Benchmarks depends on NCalsSync package, seems should be signed version
> 2. `extern alias NuGetVersion;` seems incorrect, should be
> 
> ```
> using NuGetExpression = NCalc.Expression;
> using NuGetDomain = NCalc;
> ```
> 
> 3. Solution seems to be incorrect,
> 
> ```
> {FC5C9A17-744F-4F97-B0FC-330B0514B395}.SignedRelease|Any CPU.Active Cfg = Debug|Any CPU
> {FC5C9A17-744F-4F97-B0FC-330B0514B395}.SignedRelease|Any CPU.Build.0 = Debug|Any CPU
> ```
> 
> it should be with Release

1. Using the .Signed NuGet version causes an assembly conflict with the .csproj version
2. Agree
3. Good catch

Bykiev commented:
For some reason I can't build Benchmarks project when references non-signed version and NuGet shows NCalcSync 1,0,0 installed, but the reference is correct in a project file 
`<PackageReference Include="NCalcSync" Version="3.13.1" />`

Bykiev commented:
With SignedRelease got this error:

> //    * Assembly NCalc.Benchmarks which defines benchmarks is non-optimized
> Benchmark was built without optimization enabled (most probably a DEBUG configuration). Please, build it in RELEASE.
> If you want to debug the benchmarks, please see https://benchmarkdotnet.org/articles/guides/troubleshooting.html#debugging-benchmarks.

gumbarros commented:
@Bykiev I reproduced and fixed

Bykiev commented:
Working now, but looks weired. Let's leave benchmark project as is.

gumbarros commented:
> Working now, but looks weired. Let's leave benchmark project as is.

Yeah, I will study later on how to benchmark both .csproj and NuGet, your way looked way better, but certainly was not the NuGet version

Bykiev commented:
`BinaryExpressionType.Unknown` is not used anymore, should we remove it?

gumbarros commented:
> `BinaryExpressionType.Unknown` is not used anymore, should we remove it?

I see no problem if it's not used. I think no consumer is using this enum value.

Bykiev commented on src/NCalc/ExpressionOptions.cs:
```diff
@@ -1,46 +1,50 @@
-    DecimalAsDefault = 64
-}
-
-public static class EvaluateOptionsExtensions 
```
Seems this class is not used anymore, let's remove it

gumbarros replied:
No, it's used as a extension method, like this:

```cs
bool ignoreCase = Options.HasOption(ExpressionOptions.IgnoreCase);
```

Bykiev replied:
Let's remove it then

gumbarros replied:
But the class is used, but in other name:

```cs
public static class ExpressionOptionsExtensions
{
    public static bool HasOption(this ExpressionOptions options, ExpressionOptions option)
    {
        return (options & option) == option;
    }
}
```

Bykiev replied:
So, in changelog we can add`EvaluateOptionsExtensions` renamed to` ExpressionOptionsExtensions `

Bykiev reviewed (approved):
Good job, it was really very interesting and hard to migrate from ANTLR to Parlot. The performance increased more than two times. I think it's ready for the first release. We should continue focus on performance in the next release. Comma as decimal separator is a nice to have feature, asked by many users, it'll be easier to implement it with Parlot.

yallie commented:
Awesome job, guys! 🎉
